### PR TITLE
add planning unit vector tile preview wip

### PR DESCRIPTION
**Pull request by elpamart**
## Substitute this line for a meaningful title for your changes

Planning units preview view - vector tile service. The pr contains the implementation of the endpoint on the proxy module.

Todo: need to implement the query params

### Overview

_Please write a description. If the PR is hard to understand, provide a quick explanation of the code._

### Designs

_Link to the related design prototypes (if applicable)_

### Testing instructions

_Please explain how to test the PR: ID of a dataset, steps to reach the feature, etc._

### Feature relevant tickets

_Link to the related task manager tickets_

---

## Checklist before submitting

- [ ] Meaningful commits and code rebased on `develop`.
- [ ] If this PR adds feature that should be tested for regressions when
      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
- [ ] Update CHANGELOG file

### CODE REVIEW

File: geoprocessing/test/index.html [hunk 1]
```diff
@@ -80,54 +88,65 @@
             'minzoom': 1,
             'maxzoom': 22
         },
+        // {
```
Comment (kgajowy) [inline geoprocessing/test/index.html:91]: Commented code can be removed from source. If this piece may be usable in future, we can always link to particular SHA in codebase on wiki/gist etc.

File: geoprocessing/src/modules/planning-units/planning-units.geo.entity.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  @ApiProperty()
+  @PrimaryColumn('uuid', { name: 'id' })
+  id: string;
+
+  /**
+   * geometry column.
+   */
+  @ApiProperty()
+  @Column('geometry', { name: 'the_geom', select: false })
+  theGeom: any;
```
Comment (kgajowy) [inline geoprocessing/src/modules/planning-units/planning-units.geo.entity.ts:21]: We should avoid using `any` whenever possible. 

Take a look [there](https://github.com/Vizzuality/marxan-cloud/pull/150/files#diff-e860777d8791413043fe2da6becf6ed22d1471e696f4f5a13bd78460f1d69f2eR46) for proper typing on `Geometry`

File: geoprocessing/src/modules/planning-units/planning-units.controller.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  @Header('Content-Disposition', 'attachment')
+  @Header('Access-Control-Allow-Origin', '*')
+  @Header('Content-Encoding', 'gzip')
+  async getTile(
+    @Param() tileRequest: TileRequest,
+    @Query('bbox') bbox: number[],
+    @Query('planningUnitGridShape') planningUnitGridShape: string,
+    @Query('planningUnitAreakm2') planningUnitAreakm2: number,
+    @Res() response: Response,
+  ): Promise<Object> {
```
Comment (kgajowy) [inline geoprocessing/src/modules/planning-units/planning-units.controller.ts:79]: so actually it sends `Buffer`, not `Object`, right?

File: geoprocessing/src/modules/planning-units/planning-units.controller.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  @Header('Content-Type', 'application/x-protobuf')
+  @Header('Content-Disposition', 'attachment')
+  @Header('Access-Control-Allow-Origin', '*')
+  @Header('Content-Encoding', 'gzip')
+  async getTile(
+    @Param() tileRequest: TileRequest,
+    @Query('bbox') bbox: number[],
+    @Query('planningUnitGridShape') planningUnitGridShape: string,
+    @Query('planningUnitAreakm2') planningUnitAreakm2: number,
+    @Res() response: Response,
```
Comment (kgajowy) [inline geoprocessing/src/modules/planning-units/planning-units.controller.ts:78]: Injecting `Response` is not necessary, NestJS handles it on its own, so you should be able to safely return `tile` below and remove `@Res()`

File: geoprocessing/src/modules/planning-units/planning-units.controller.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  @ApiBadRequestResponse()
+  @Header('Content-Type', 'application/x-protobuf')
+  @Header('Content-Disposition', 'attachment')
+  @Header('Access-Control-Allow-Origin', '*')
+  @Header('Content-Encoding', 'gzip')
+  async getTile(
+    @Param() tileRequest: TileRequest,
+    @Query('bbox') bbox: number[],
+    @Query('planningUnitGridShape') planningUnitGridShape: string,
+    @Query('planningUnitAreakm2') planningUnitAreakm2: number,
```
Comment (kgajowy) [inline geoprocessing/src/modules/planning-units/planning-units.controller.ts:77]: It is possible to create a `dto` file with all those definitions and have Nest include wanted params for you:
`@Query() inputs: YourDtoClass`.
You could then also move Api-Doc decorators to that `dto` file.

On top of that - currently there is no validation so you can receive `bbox` as `string` or `null` and so on.

File: geoprocessing/src/modules/planning-units/planning-units.controller.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  ApiParam,
+  ApiQuery,
+  ApiBadRequestResponse,
+} from '@nestjs/swagger';
+import { TileRequest } from 'src/modules/tile/tile.service';
+
+import { Response } from 'express';
+
+@Controller(`${apiGlobalPrefixes.v1}`)
+export class PlanningUnitsController<T> {
```
Comment (kgajowy) [inline geoprocessing/src/modules/planning-units/planning-units.controller.ts:23]: Wondering if there is reason we need generic type there?

File: geoprocessing/src/modules/planning-units/planning-units.controller.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  private readonly logger: Logger = new Logger(PlanningUnitsController.name);
```
Comment (kgajowy) [inline geoprocessing/src/modules/planning-units/planning-units.controller.ts:24]: Seems to be not used anywhere.

File: api/src/modules/proxy/proxy.controller.ts [hunk 8]
```diff
@@ -161,4 +161,60 @@ export class ProxyController {
...
+    type: String,
+    required: false,
+  })
+  @ApiQuery({
+    name: 'planningUnitAreakm2',
+    description: 'Planning unit area in km2',
+    type: Number,
+    required: false,
+  })
+  @Get('/planning-units/preview/tiles/:z/:x/:y.mvt')
```
Comment (hotzevzl) [inline api/src/modules/proxy/proxy.controller.ts:213]: may I suggest to start moving these endpoints to the controller they logically belong to?

In practice, we should get rid of this ProxyController, after all, I think - unless we have proxied requests that don't belong elsewhere. So this specific new endpoint in this PR could be moved to a (new in this case) `PlanningUnitsController` within the (existing) `PlanningUnitsModule`.

If you agree with moving these endpoints, feel free to do this in a separate PR, of course.

The main benefit from my point of view is that we would be grouping them together with related endpoints (so the features tiles endpoint could go to the existing GeoFeaturesController, and so on), which also helps with the generated Swagger documentation.

**Comment by hotzevzl** [comment]
Superceded by https://github.com/Vizzuality/marxan-cloud/pull/182

**Pull request closed without merging**